src/graph/TruncateSpaceExecutor.cpp

```diff
@@ -0,0 +1,199 @@
+    auto future = ectx()->getMetaClient()->renameSpace(tempSpaceName_, *spaceName_);
+    auto *runner = ectx()->rctx()->runner();
+
+    auto cb = [this] (auto &&resp) {
+        if (!resp.ok()) {
+            onError_(std::move(resp).status());
+            return;
+        }
+        auto  ret = std::move(resp).value();
+        if (!ret) {
+            LOG(ERROR) << "Rename space " << tempSpaceName_ << " failed";
+            auto outFmt = "Rename space `%s' failed. Your need to "
+                          "rename the space `%s' to `%s' by yourself when the metad is OK";
+            auto error = folly::stringPrintf(outFmt, spaceName_->c_str(),
+                    tempSpaceName_.c_str(), spaceName_->c_str());
+            onError_(Status::Error(error));
+            return;
+        }
+
+        // If current space is the truncate space, need to update the spaceId in session
+        auto useSpaceId = ectx()->rctx()->session()->space();
+        if (useSpaceId > 0 && fromSpaceId_ == useSpaceId) {
+            VLOG(1) << "Update current spaceId to " << newSpaceId_;
+            ectx()->rctx()->session()->setSpace(*spaceName_, newSpaceId_);
+        }
+
+        onFinish_(Executor::ProcessControl::kNext);
+    };
+
+    auto error = [this] (auto &&e) {
+        auto msg = folly::stringPrintf("Rename space `%s' exception: %s",
+                                       tempSpaceName_.c_str(), e.what().c_str());
+        LOG(ERROR) << msg;
+        onError_(Status::Error(std::move(msg)));
+        return;
+    };
+
+    std::move(future).via(runner).thenValue(cb).thenError(error);
+}
+
+void TruncateSpaceExecutor::dropSpace(const std::string &spaceName, bool isIfExists) {
+    auto future = ectx()->getMetaClient()->dropSpace(spaceName, isIfExists);
+    auto *runner = ectx()->rctx()->runner();
+
+    auto cb = [spaceName, isIfExists, this] (auto &&resp) {
+        // If drop from_space failed, need to drop the temp_space
+        if (!resp.ok()) {
+            if (!isIfExists) {
+                dropSpace(tempSpaceName_, true);
+            }
+            onError_(std::move(resp).status());
+            return;
+        }
+        auto ret = std::move(resp).value();
+        if (!ret) {
+            if (!isIfExists) {
+                dropSpace(tempSpaceName_, true);
+            }
+            onError_(Status::Error("Drop space `%s' failed.", spaceName.c_str()));
+            return;
+        }
+
+        if (!isIfExists) {
+            renameSpace();
+        } else {
+            onError_(Status::Error("Drop space `%s' failed.", spaceName.c_str()));
+        }
+        return;
+    };
+
+    auto error = [spaceName, this] (auto &&e) {
+        auto msg = folly::stringPrintf("Drop space `%s' exception: %s",
+                                       spaceName.c_str(), e.what().c_str());
+        LOG(ERROR) << msg;
+        dropSpace(tempSpaceName_, true);
+        onError_(Status::Error(std::move(msg)));
```

Comment:
dropSpace is a asyc interface, here can not call onError_ directly

Reply:
I will delete it.